text/000-rocq-findlib.md

```diff
@@ -0,0 +1,267 @@
+```
+ocamlfind query some_dep -format "%(rocqpath)"
+```
+The result is empty if `some_dep` is a standard OCaml package (possibly a Rocq
+plugin), and it gives a directory path for Rocq theories.
+
+The mapping for each theory package `some_theory` can then be computed using
+the following command.
+```
+ocamlfind query some_theory -format "-Q %d/rocq.d %(rocqpath)"
+```
+
+## Example: stdpp
+
+The packages for [stdpp](https://gitlab.mpi-sws.org/iris/stdpp) could end up
+being installed under `<PREFIX>/lib/rocq-stdpp/`, which would contains the
+following `META` file.
+```
+version = "dev"
+description = "An extended \"Standard Library\" for Rocq."
+rocqpath = "stdpp"
+requires = "rocq-stdlib"
+
+package "bitvector" (
+  directory = "bitvector"
+  version = "dev"
+  description = "A library for bitvectors based on std++."
+  rocqpath = "stdpp.bitvector"
+  requires = "rocq-stdpp"
+)
+
+package "unstable" (
+  directory = "unstable"
+  version = "dev"
+  description = "Unfinished std++ libraries."
+  rocqpath = "stdpp.unstable"
+  requires = "rocq-stdpp"
+)
+```
+
+Differences with the current `stdpp` setup:
+- Rocq-ified package names.
+- Use of sub-packages form `stdpp.bitvector` and `stdpp.unstable`, instead of
+  different Opam packages.
+
+## Example: coq-elpi
+
+```
+rocqpath = "elpi"
+requires = "rocq-stdlib coq-elpi.plugin"
+...
+
+package "plugin" (
+  directory = "plugin"
+  requires = "rocq-core.plugins.ltac rocq-core.vernac elpi"
+  ...
+)
+
+package "apps" (
+  directory "apps"
+
+  package "derive" (
+    directory "derive"
+    rocqpath = "elpi.apps.derive"
+    requires = "rocq-elpi rocq-elpi.apps.derive.elpi"
+    ...
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.derive.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  package "tc" (
+    directory "tc"
+    rocqpath = "elpi.apps.tc"
+    requires = "rocq-elpi rocq-elpi.apps.tc.elpi rocq-elpi.apps.tc.plugin"
+    ...
+
+    package "plugin" (
+      directory = "plugin"
+      requires = "rocq-core.plugins.ltac rocq-core.vernac rocq-elpi.plugin"
+      ...
+    )
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.tc.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  ...
+)
+```
+
+# Drawbacks
+
+None, except the eventual need to port all projects to use a build system that
+respects the new installation scheme. However, as a first step, it is possible
+to keep using the current installation scheme at the same time.
+
+# Alternatives
+
+All reasonable systems have a modular installation scheme. An alternative
+implementation would be to start from scratch and not reuse the Findlib file
+format, but given the close integration of Rocq in the OCaml ecosystem that
+probably would not make sense.
+
+# Unresolved questions
+
+Can we adapt `rocq_makefile` so that it expects dependencies to be installed
+following the considered installation scheme, and itself installs libraries
+appropriately?
```

Comment:
Having `rocq_makefile` install packages may be nice as an option.  However, developers modifying a package should probably still be able to build it without installing anything (instead reporting missing packages as error).

If you're modifying a package, the code you're modifying should not be in opam or Rocq directories (good development practice).  There's also the more obscure (hypothetical?) case where you're modifying multiple packages at once--all of which should be outside of opam/Rocq directories.

One difficulty with modifying `rocq_makefile` is that IINM GUIs know and parse the file.  (A better design would be for only Rocq to parse the file and provide an API the GUIs can call to get the details in a processed format.)

Reply:
Currently, the typical workflow for maintaining packages is to create an opam switch with all dependencies installed in the appropriate version, and then building the project from its sources. For example, I know that this is how most Iris devs do things, and they even use some infrastructure for pinning the right versions of their dependencies in opam.

Alternatives exist, like using `dune` and it's composed builds, which lets you compile a project and its dependencies in the same workspace, without having to install said dependencies. What you are talking about above seems to be similar to that if I understand correction. In any case, a mechanism like that is very useful, but rather orthogonal to the particular installation scheme / package metadata.

I have never heard of anything like GUIs parsing `rocq_makefile` files. Do you have pointers to that?

Reply:
> orthogonal to the particular installation scheme / package metadata.

My first paragraph above responds directly to your comments about installing libraries.  Also what I said about using a non-opam, non-Rocq directory for packages that you're modifying.  That is directly related to the installation scheme, isn't it? 

> Currently, the typical workflow for maintaining packages is to create an opam switch

Perhaps, but I expect that requires considerable familiarity with opam.  The procedure should be documented rather than forcing everyone to figure it out on their own.  Last I checked, about 2 years ago, the opam doc was still quite incomplete.  Pretty much the same with the dune documentation and its dependencies.  Does this procedure let you put the code you want to modify in a non-opam directory?

I wrote [this section](https://coq.inria.fr/doc/V8.20.0/refman/practical-tools/utilities.html#coq-configuration-basics) about 2 years ago.  Some of the procedures are quite awkward due to missing features.

> I have never heard of anything like GUIs parsing `rocq_makefile` files. Do you have pointers to that?

The dependency is on `_CoqProject`, my mistake.

Reply:
> > orthogonal to the particular installation scheme / package metadata.
> 
> My first paragraph above responds directly to your comments about installing libraries. Also what I said about using a non-opam, non-Rocq directory for packages that you're modifying. That is directly related to the installation scheme, isn't it?

I'm sorry, I realize now that I did not quite understand what you meant in your first paragraph, and in fact I still don't. Clearly, I'm failing to communicate what I have in mind. What kind of situation do you have in mind exactly when writing "developers modifying a package"?

In case that helps, let me try to clarify what I meant to communicate in the CEP paragraph we are discussing: I have not yet worked out how `rocq_makefile` can be made to work with the discussed installation scheme. That roughly means figuring out two things:
1. How does a developer communicate to `rocq_makefile` the theory dependencies of the project in practice? Currently there is no need to do so because everything that is installed is always available, but that will eventually need to change when we stop putting theories in `user-contrib`. Instead, we will need to be explicit about what packages we depend on, and `-Q` bindings will need to be somehow added for all transitive deps.
2. How do we make it so that `rocq_makefile` produces a `Makefile` that, when invoked with `make install`, has the effect of installing the theories as expected for the new installation scheme?

Reply:
> What kind of situation do you have in mind exactly when writing "developers modifying a package"?

Primarily the author or a maintainer of the package who needs to modify the code in the package while doing maintenance, perhaps preparing PRs, but potentially could also be a developer who wants to tweak the package locally, e.g. to work around a limitation.  Is that specific enough?